[ci] Add coq-tools to the CI

JasonGross opened the pull request:
After #12023 broke the bug minimizer, I'd like to add [coq-tools](https://github.com/JasonGross/coq-tools/) to the CI.  It's relatively light-weight (under 5 minutes, I believe), and I'd like to know when it's going to break on master before it's broken, rather than after.  It tests a relatively under-tested part of Coq, mostly (the display output of error message, by and large), and I'm happy to take responsibility for fixing it when some PR is going to break it (mainly I just want a sort-of early warning system, and I want PRs to not accidentally break it by changing things that they don't realize they're changing).

**Kind:** infrastructure.

JasonGross commented:
Ah, this will probably have to wait until coq-tools migrates from python2 to python3, I guess.

JasonGross commented:
Ah, no, apparently it works fine.  But should we wait until I migrate to python3?

Zimmi48 commented:
I'll merge this tomorrow.

JasonGross commented:
> > I'm happy to take responsibility for fixing it when some PR is going to break it
> 
> To me this is a key seller (we cannot assume to maintain Python code ourselves). I wonder how to make this clear to anyone encountering such failure...

Perhaps we can add to the CI script `|| { RV=$?; echo "The build broke, if an overlay is needed, mention @JasonGross in describing the expected change in Coq that needs to be taken into account, and he'll prepare a fix for coq-tools"; exit $RV }`?

ejgallego commented:
Obviously, I support @JasonGross request for Coq to test for regressions on the widely and really nice set of tools he has developed.

On the other hand, as I've expressed a few times in the past, I remain concerned about the interfaces these tools are using; it would be great if we could improve this part or at least have a plan.

For example I have hopes for a Coq Python interface to appear in 2020, regarding timing we will adapt the flame format widely used, etc...

ejgallego commented:
> Perhaps we can add to the CI script `|| { RV=$?; echo "The build broke, if an overlay is needed, mention @JasonGross in describing the expected change in Coq that needs to be taken into account, and he'll prepare a fix for coq-tools"; exit $RV }`?

It seems like a nice idea, maybe we should add some kind of instructions for all 3rd party devs indeed?

JasonGross commented:
> On the other hand, as I've expressed a few times in the past, I remain concerned about the interfaces these tools are using; it would be great if we could improve this part or at least have a plan.
> 
> For example I have hopes for a Coq Python interface to appear in 2020, regarding timing we will adapt the flame format widely used, etc...

Indeed, the tools currently parse error output on the command line (I think this is mostly fine), use glob file information to find the absolute names of some constants, and perhaps most worryingly, interact with `coqtop -emacs -verbose`, IIRC, to split the document into sentences and definition-proof-chunks; having a python interface exposing this information would be great (though I'd probably keep the `coqtop -emacs` interaction for older versions of Coq, as the tools are supporting all the way back to Coq 8.5 (I dropped 8.4 support when I had trouble getting launchpad packages to work correctly)).

Zimmi48 commented:
@JasonGross I'm not sure I get what the labels are for. Emilio's comment is a medium-term concern but shouldn't prevent merging this PR.

JasonGross commented:
Labels are because I need to dig into why coq-tools fails on Coq's CI but succeeds on mine.  I think it's something to do with it defaulting to ASCII rather than utf8 somewhere based on different environment variables...

JasonGross commented:
CI now passes!

Zimmi48 commented:
@JasonGross I've got a spurious failure here: https://gitlab.com/Zimmi48/coq/-/jobs/532078775
Can you look into it? I'd like to avoid having a new job that fails randomly.

JasonGross commented:
It looks like `touch ex22.v; coq_makefile -Q . Top ex22.v -o Makefile; make ex22.glob` is producing a .glob file with an older timestamp than the corresponding .v file.  Let me try adding a sleep before and after the touch, and see if that fixes things.

Zimmi48 commented:
That seems a bit fragile...

JasonGross commented:
So, it turns out there was already some code in the python script that's supposed to delay calling `make` until the timestamp of the .v file is firmly not the present moment.  I'm not sure why it wasn't sufficient, so I've just added some extra debugging information to print out the timestamps, let's see if it comes up again.

Zimmi48 commented:
New one: https://gitlab.com/Zimmi48/coq/-/jobs/535005442

JasonGross commented:
Okay, I've tried extending the delay time by a second, to hopefully not be subject to boundary conditions, and added a bit more verbosity.  This should hopefully fix things, but if it does not, please report back here on future failures.  (And thanks!)

Zimmi48 commented:
New failure that looks spurious to me: https://gitlab.com/coq/coq/-/jobs/544269051

JasonGross commented:
I'm confused why subprocess/Popen would sometimes give me stdout as a str and sometimes as bytes (this mainly matters because I'm trying to be compatible with both python 2 and 3), but I can make it robust to this.  Thanks for the report!

JasonGross commented:
Should be fixed now

Zimmi48 commented:
New failure (it had been a while): https://gitlab.com/coq/coq/-/jobs/583014463

JasonGross commented:
This is very strange.  It claims that Coq failed to compile a small fragment of code that should be compileable.  I'll add some debugging output later today so that we can see what the claimed error message is from Coq if it fails again.

JasonGross commented:
Okay, I've added more debugging info in https://github.com/JasonGross/coq-tools/commit/1abb715efff47a50f3528ef55877c47d3887b8fc , hopefully this will let us know what's going on.  Let me know if it turns up again, and thanks for the report @Zimmi48 .  My current guess is that it was a timeout, i.e., it took `coqc -q -nois`  longer than a full second to process the code
```coq
Declare ML Module "ltac_plugin".
Inductive False := .
Axiom proof_admitted : False.
Tactic Notation "admit" := abstract case proof_admitted.
```

Zimmi48 commented:
This is probably another spurious failure: https://gitlab.com/coq/coq/-/jobs/591578724

JasonGross commented:
One of these was a timeout where the time taken by one Coq run was more than 2x the time taken by the previous one.  I've increased the fudge factor in allowing for timeouts, hopefully this fixes things.    I'm rather confused by the other error, will try to dig into it later

JasonGross commented:
I'm extremely confused by the non-timeout error: it seems that somehow either python failed to create a file (or an unexpected error happened very early in the process), or else python returned the wrong answer from the OS about whether or not a file exists.  I've inserted a bit more debugging output into this particular example so I can debug a bit more if this happens again.

Zimmi48 commented:
Again an error: https://gitlab.com/coq/coq/-/jobs/592111000

JasonGross commented:
That's a transient error, sorry.  I forgot to update the tests when I updated the timeout, and realized this when my CI failed and fixed it.  This build must have been run in the intervening time.  If you restart it, it should succeed
